agent/testagent.go

```diff
@@ -184,18 +204,36 @@ func (a *TestAgent) Start(t *testing.T) *TestAgent {
 		// the data dir, such as in the Raft configuration.
 		if a.DataDir != "" {
 			if err := os.RemoveAll(a.DataDir); err != nil {
-				require.Fail("%s %s Error resetting data dir: %s", id, a.Name, err)
+				return fmt.Errorf("%s %s Error resetting data dir: %s", id, a.Name, err)
 			}
 		}
 	}
 
 	// Start the anti-entropy syncer
 	a.Agent.StartSync()
 
+	err = a.waitForUp()
+	if err != nil {
+		return err
+	}
+
+	a.dns = a.dnsServers[0]
+	a.srv = a.httpServers[0]
+	return nil
+}
+
+// waitForUp waits for leader election, or waits for the agent HTTP
+// endpoint to start responding, depending on the agent config.
+func (a *TestAgent) waitForUp() error {
+	timer := retry.TwoSeconds()
+	deadline := time.Now().Add(timer.Timeout)
+
+	var retErr error
 	var out structs.IndexedNodes
-	retry.Run(t, func(r *retry.R) {
+	for !time.Now().After(deadline) {
 		if len(a.httpServers) == 0 {
-			r.Fatal(a.Name, "waiting for server")
+			retErr = fmt.Errorf("%s: waiting for server", a.Name)
+			continue // fail, try again
```

Comment:
All the `continue` statements here seem like they will avoid the `time.Sleep` at the end of the loop and burn cpu cycles. Should that sleep be moved to the beginning of the loop?

Alternatively you could wrap all of the checks in for loop like so and break instead of continue:

```go
for !time.Now().After(deadline) {
   for {
     if <check> {
       retErr = ...
       break
     }
   }
   
   time.Sleep(timer.Wait)
}
```

That way the sleep can be avoided if everything is already ready but we can still prevent the hot loop.

Reply:
oh duh! thanks matt, great catch. i'll add it to the loop header.